### PR TITLE
Seed is set for deterministic results in development

### PR DESCRIPTION
**Pull request by davidwyld**
I've set the seeds at the start of `main.py` for `numpy.random.seed` and `random.seed` as detailed in issue RSE-Sheffield/deepquest/issues/4

Using the test dataset generated by `getTestData.py` this gets the following results running with sentence level, BiRNN, on cpu, with patience=5 if you want to try and reproduce them:
*  `theano 1.0`: PCC=0.19076 @ epoch3, running until epoch 8.
* `TensorFlow 1.13.12`: PCC = 0.163548 @ epoch3, running until epoch 8.

I haven't tested this on gpu yet, as described here:
* https://machinelearningmastery.com/reproducible-results-neural-networks-keras/
and here:
* https://keras.io/getting-started/faq/#how-can-i-obtain-reproducible-results-using-keras-during-development
this determinism may break on gpu but we can do things to get around that at least for development purposes.

At present, the `RND_SEED` variable is not used. But we can add that in during a later revision if needs be. We may want to change this, because arguably we only want this kind of determinism for testing.

### Also, TensorFlow
Supposedly (according to the above links) TensorFlow's seed needs to be set as well, but I haven't implemented this at present and it is giving stable results in this test case.
TensorFlow's seed would be set as follows:

```python
from tensorflow import set_random_seed
set_random_seed(2)
```

### CODE REVIEW

**Comment by davidwyld** [comment]
@fredblain I can't choose you as a referee (because you're outside the organization?) but I guess you'll want to see this too.

**Comment by davidwyld** [comment]
**ALSO** I've modified my local version of `train-test-sentQEbRNN.sh` so that when python is called on lines 95 & 113, `PYTHONHASHSEED` is set to `0`. As follows:
```bash
THEANO_FLAGS=device=$device PYTHONHASHSEED=0 python main.py .....
```
we will need to integrate this in however deepquest is ultimately run.

**Comment by drj11** [comment]
I've run this branch on my laptop with the following diff made to `config.py`:

```
diff --git a/config.py b/config.py
index f79e1ea..59ce7e5 100644
--- a/config.py
+++ b/config.py
@@ -5,7 +5,7 @@ def load_parameters():
     """
 
     # Input data params
-    TASK_NAME = 'qe-2016'                           # Task name
+    TASK_NAME = 'wmt15'                           # Task name
     DATASET_NAME = TASK_NAME                        # Dataset name
     SRC_LAN = 'src'                                  # Language of the source text
     TRG_LAN = 'mt'                                  # Language of the target text
```

(that is, just changing the `TASK_NAME`)

using the command line:

```
env PYTHONHASHSEED=0 time python main.py
```

I can reliably get the same results on my own machine, but they seem to differ from the numbers that @davidwilby reports in this PR.

My final line is:

```
I1008 11:43:01.524625 140057880614656 callbacks.py:974] ---epoch 11: early stopping. Best pearson found at epoch 6: 0.191172
```

File: main.py [hunk 1]
```diff
@@ -1,7 +1,13 @@
 # -*- coding: utf-8 -*-
-
 from __future__ import print_function
 
+from numpy.random import seed
+seed(1) # this is a test, not using RND_SEED variable here
```
Comment (drj11) [inline main.py]: Remove the comment, as it looks like it's a "note to self" type comment made during development.

**Review by drj11 — APPROVED** [review]
address the only comment, then you can merge at will. Or I can.

**Comment by drj11** [comment]
The fact that we get different numbers reported is probably because we're using different configs and/or the shell scripts are also settings things. Which is a problem that we need to investigate, but we can do that separately.

**Comment by drj11** [comment]
My reported config is:
```
	ADDITIONAL_OUTPUT_MERGE_MODE: Add
	ALIGN_FROM_RAW: True
	ALPHA_FACTOR: 0.6
	APPLY_DETOKENIZATION: False
	ATTENTION_SIZE: 500
	BATCH_NORMALIZATION_MODE: 1
	BATCH_SIZE: 50
	BEAM_SEARCH: False
	BEAM_SIZE: 6
	BIDIRECTIONAL_DEEP_ENCODER: True
	BIDIRECTIONAL_ENCODER: True
	BPE_CODES_PATH: examples/wmt15//training_codes.joint
	CLASSIFIER_ACTIVATION: softmax
	CLIP_C: 1.0
	CLIP_V: 0.0
	COVERAGE_NORM_FACTOR: 0.2
	COVERAGE_PENALTY: False
	DATASET_NAME: wmt15
	DATASET_STORE_PATH: datasets/
	DATA_AUGMENTATION: False
	DATA_ROOT_PATH: examples/wmt15/
	DECODER_HIDDEN_SIZE: 500
	DECODER_RNN_TYPE: ConditionalGRU
	DEEP_OUTPUT_LAYERS: [('linear', 300)]
	DETOKENIZATION_METHOD: detokenize_none
	DOUBLE_STOCHASTIC_ATTENTION_REG: 0.0
	DROPOUT_P: 0.0
	EARLY_STOP: True
	ENCODER_HIDDEN_SIZE: 50
	ENCODER_RNN_TYPE: GRU
	EPOCHS_FOR_SAVE: 1
	EPOCH_PER_EST_SENT: 10
	EPOCH_PER_EST_WORD: 10
	EPOCH_PER_PRED: 5
	EPOCH_PER_UPDATE: 1
	EVAL_EACH: 1
	EVAL_EACH_EPOCHS: True
	EVAL_ON_SETS: ['val', 'test']
	EVAL_ON_SETS_KERAS: []
	EXTRA_NAME: 
	FILL: end
	FORCE_RELOAD_VOCABULARY: False
	HEURISTIC: 0
	HOMOGENEOUS_BATCHES: False
	INIT_ATT: glorot_uniform
	INIT_FUNCTION: glorot_uniform
	INIT_LAYERS: ['tanh']
	INNER_INIT: orthogonal
	INPUTS_IDS_DATASET: ['source_text', 'target_text']
	INPUTS_IDS_MODEL: ['source_text', 'target_text']
	INPUT_VOCABULARY_SIZE: 3224
	JOINT_BATCHES: 4
	LENGTH_NORM_FACTOR: 0.2
	LENGTH_PENALTY: False
	LOSS: ['mse']
	LR: 1.0
	LR_DECAY: None
	LR_GAMMA: 0.8
	LR_HALF_LIFE: 5000
	LR_REDUCER_EXP_BASE: 0.5
	LR_REDUCER_TYPE: exponential
	LR_REDUCE_EACH_EPOCHS: False
	LR_START_REDUCTION_ON_EPOCH: 0
	MAPPING: examples/wmt15//mapping.src_mt.pkl
	MAXLEN_GIVEN_X: True
	MAXLEN_GIVEN_X_FACTOR: 2
	MAX_EPOCH: 500
	MAX_INPUT_TEXT_LEN: 70
	MAX_OUTPUT_TEXT_LEN: 70
	MAX_OUTPUT_TEXT_LEN_TEST: 210
	MAX_SRC_INPUT_TEXT_LEN: 70
	MAX_TRG_INPUT_TEXT_LEN: 70
	METRICS: ['qe_metrics']
	MINLEN_GIVEN_X: True
	MINLEN_GIVEN_X_FACTOR: 3
	MIN_OCCURRENCES_INPUT_VOCAB: 0
	MIN_OCCURRENCES_OUTPUT_VOCAB: 0
	MODE: training
	MODEL_NAME: wmt15_srcmt_EncSent
	MODEL_TYPE: EncSent
	MULTI_TASK: False
	NOISE_AMOUNT: 0.01
	NORMALIZE_SAMPLING: False
	NO_REF: False
	N_LAYERS_DECODER: 1
	N_LAYERS_ENCODER: 1
	OPTIMIZED_SEARCH: True
	OPTIMIZER: Adadelta
	OUTPUTS_IDS_DATASET: ['sent_qe']
	OUTPUTS_IDS_MODEL: ['sent_qe']
	OUTPUT_VOCABULARY_SIZE: 7462
	PAD_ON_BATCH: False
	PARALLEL_LOADERS: 1
	PATIENCE: 5
	POS_UNK: True
	PRED_SCORE: hter
	QE_VECTOR_SIZE: 75
	REBUILD_DATASET: True
	RECURRENT_DROPOUT_P: 0.0
	RECURRENT_INPUT_DROPOUT_P: 0.0
	RECURRENT_WEIGHT_DECAY: 0.0
	RELOAD: 0
	RELOAD_EPOCH: True
	SAMPLE_WEIGHTS: {'word_qe': {'BAD': 3}}
	SAMPLING: max_likelihood
	SAMPLING_SAVE_MODE: list
	SAVE_EACH_EVALUATION: True
	SEARCH_PRUNING: False
	SKIP_VECTORS_HIDDEN_SIZE: 300
	SOURCE_TEXT_EMBEDDING_SIZE: 300
	SRC_LAN: src
	SRC_PRETRAINED_VECTORS: None
	SRC_PRETRAINED_VECTORS_TRAINABLE: True
	START_EVAL_ON_EPOCH: 1
	STOP_METRIC: pearson
	STORE_PATH: trained_models/wmt15_srcmt_EncSent/
	TARGET_TEXT_EMBEDDING_SIZE: 300
	TASK_NAME: wmt15
	TEMPERATURE: 1
	TEXT_FILES: {'train': 'train.', 'val': 'dev.', 'test': 'test.'}
	TOKENIZATION_METHOD: tokenize_none
	TOKENIZE_HYPOTHESES: True
	TOKENIZE_REFERENCES: True
	TRAIN_ON_TRAINVAL: False
	TRG_LAN: mt
	TRG_PRETRAINED_VECTORS: None
	TRG_PRETRAINED_VECTORS_TRAINABLE: True
	USE_BATCH_NORMALIZATION: True
	USE_L2: False
	USE_NOISE: True
	USE_PRELU: False
	VERBOSE: 1
	WEIGHT_DECAY: 0.0001
	WORD_QE_CLASSES: 5
	WRITE_VALID_SAMPLES: True
```

File: main.py (hunk 1 above)
Comment (davidwyld) [reply main.py]: Done: https://github.com/RSE-Sheffield/nmt-keras/commit/3909db582df38dfbc30c7158485ac5333538099e

**Comment by drj11** [comment]
Using the shell script, I can reproduce the Pearson Correlation Coefficient that you get.
From my `log-wmt15_srcmt_EncSent.txt` file:

```
---epoch 8: early stopping. Best pearson found at epoch 3: 0.191076
```

**Comment by fredblain** [comment]
I would suggest to move all global variables, seeds, etc. to a common file (e.g. under a `/lib/utils.py`) so they can easily be called, regardless of how the end user is running the toolkit (as a module or via command line).
This could be done at the end, but considering it now would ease the refactoring work.
